src/GafferUI/StandardNodeGadget.cpp

```diff
@@ -1222,14 +1263,32 @@ bool StandardNodeGadget::drop( GadgetPtr gadget, const DragDropEvent &event )
 	connect( event, m_dragDestination );
 
 	m_dragDestination->setHighlighted( false );
-	for( StandardNodule::RecursiveIterator it( this ); !it.done(); ++it )
-	{
-		(*it)->setLabelVisible( false );
-	}
+	updatePersistentNoduleLabels();
+
 	m_dragDestination = nullptr;
 	return true;
 }
 
+void StandardNodeGadget::noduleAdded( Nodule *nodule )
+{
+	if( auto standardNodule = IECore::runTimeCast<StandardNodule>( nodule ) )
+	{
+		bool inputVisible = false;
+		bool outputVisible = false;
+
+		if( IECore::ConstBoolDataPtr d = Gaffer::Metadata::value<IECore::BoolData>( node(), g_inputNoduleLabelsVisibleKey ) )
+		{
+			inputVisible = d->readable();
+		}
+		if( IECore::ConstBoolDataPtr d = Gaffer::Metadata::value<IECore::BoolData>( node(), g_outputNoduleLabelsVisibleKey ) )
+		{
+			outputVisible = d->readable();
+		}
```

Comment:
We could test the plug direction first here, and then only look up the single bit of relevant metadata.

Reply:
Fixed in https://github.com/GafferHQ/gaffer/pull/6135/commits/31bf3651b2f9316f7fbc80631ed93f7f51c53415